File: app/src/main/java/br/com/appinfra/appinfra/models/FirebaseServices/FirebaseHelper.java
package br.com.appinfra.appinfra.models.FirebaseServices;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseException;
import com.google.firebase.database.DatabaseReference;

import java.util.ArrayList;

import br.com.appinfra.appinfra.models.beans.Complaint;

/**
 * Created by alber on 17/11/2017.
 */

public class FirebaseHelper {

        DatabaseReference db;
        Boolean saved=null;
        ArrayList<Complaint> Complaints=new ArrayList<>();
        /*
         PASS DATABASE REFRENCE
     */
        public FirebaseHelper(DatabaseReference db) {
            this.db = db;
        }
        //WRITE IF NOT NULL
        public Boolean save(Complaint Complaint)
        {
            if(Complaint==null)
            {
                saved=false;
            }else
            {
                try
                {
                    db.child("list_complaints").push().setValue(Complaint);
                    saved=true;
                }catch (DatabaseException e)
                {
                    e.printStackTrace();
                    saved=false;
                }
            }
            return saved;
        }
        //IMPLEMENT FETCH DATA AND FILL ARRAYLIST
        private void fetchData(DataSnapshot dataSnapshot)
        {
            Complaints.clear();
            for (DataSnapshot ds : dataSnapshot.getChildren())
            {
                Complaint Complaint = ds.getValue(Complaint.class);
                Complaints.add(Complaint);
            }
        }
        //READ BY HOOKING ONTO DATABASE OPERATION CALLBACKS
        public ArrayList<Complaint> retrieve()
        {
            db.addChildEventListener(new ChildEventListener() {
                @Override
                public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                    fetchData(dataSnapshot);
                }
                @Override
                public void onChildChanged(DataSnapshot dataSnapshot, String s) {
                    fetchData(dataSnapshot);
                }
                @Override
                public void onChildRemoved(DataSnapshot dataSnapshot) {
                }
                @Override
                public void onChildMoved(DataSnapshot dataSnapshot, String s) {
                }
                @Override
                public void onCancelled(DatabaseError databaseError) {
                }
            });
            return Complaints;
        }

    }

File: app/src/main/java/br/com/appinfra/appinfra/models/FirebaseServices/FirebaseHelper.kt
package br.com.appinfra.appinfra.models.FirebaseServices

import br.com.appinfra.appinfra.models.beans.Complaint
import com.google.firebase.database.*
import java.util.*

/**
 * Created by alber on 17/11/2017.
 */

// Class Functions Complaints Firebase
class FirebaseHelper(internal var db: DatabaseReference) {

    internal var saved: Boolean? = null
    internal var Complaints = ArrayList<Complaint>()

    // Save Complaint
    fun save(Complaint: Complaint?): Boolean? {
        if (Complaint == null) {
            saved = false
        } else {
            try {
                db.child("list_complaints").push().setValue(Complaint)
                saved = true
            } catch (e: DatabaseException) {
                e.printStackTrace()
                saved = false
            }
        }
        return saved
    }

    // Read Complaints in Firebase
    fun retrieve(): ArrayList<Complaint> {

        db.addChildEventListener(object : ChildEventListener {

            override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {
                Complaints.clear()
                for (ds in dataSnapshot.children) {
                    val Complaint = ds.getValue(Complaint::class.java)
                    if (Complaint != null) {
                        Complaints.add(0,Complaint)
                    }
                }
            }

            override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {
                Complaints.clear()
                for (ds in dataSnapshot.children) {
                    val Complaint = ds.getValue(Complaint::class.java)
                    if (Complaint != null) {
                        Complaints.add(Complaint)
                    }
                }
            }

            override fun onChildRemoved(dataSnapshot: DataSnapshot) {
                Complaints.clear()
                for (ds in dataSnapshot.children) {
                    val Complaint = ds.getValue(Complaint::class.java)
                    if (Complaint != null) {
                        Complaints.add(Complaint)
                    }
                }
            }
            override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {
                Complaints.clear()
                for (ds in dataSnapshot.children) {
                    val Complaint = ds.getValue(Complaint::class.java)
                    if (Complaint != null) {
                        Complaints.add(Complaint)
                    }
                }
            }
            override fun onCancelled(databaseError: DatabaseError) {}
        })

        return Complaints
    }



}
